### PR TITLE
[Tensor] Add tensor cat

### PR DESCRIPTION
**Pull request by zhoonit**
## Commits to be reviewed in this PR

<details><summary>[tensor] Add tensor::split</summary><br />

This patch add tensor split for wider use

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>

</details>

<details><summary>[Tensor] Add tensor cat</summary><br />

This patch add tensor cat method and it's test

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>

</details>

### CODE REVIEW

File: nntrainer/tensor/tensor.cpp [hunk 1]
```diff
@@ -750,6 +752,49 @@ Tensor Tensor::getSharedDataTensor(const TensorDim dim_, unsigned int offset,
...
+      }
+      break;
+    }
+    return value;
+  };
+
+  std::vector<Tensor> ret;
+  ret.reserve(num_size);
+
+  for (unsigned int i = 0; i < num_size; ++i) {
```
Comment (kparichay) [inline nntrainer/tensor/tensor.cpp:787]: I would highly recommend you to refer the split layer implementation, where changes the shape the tensor to be split to not only simplify the computation but also makes it quite efficient.

File: nntrainer/tensor/tensor.cpp [hunk 2]
```diff
@@ -750,6 +752,49 @@ Tensor Tensor::getSharedDataTensor(const TensorDim dim_, unsigned int offset,
   return ret;
 }
 
+std::vector<Tensor> Tensor::split(unsigned num_size, int axis) {
+  if (axis == -1) {
+    axis = 3;
+  }
+  NNTR_THROW_IF(!(0 <= axis && axis < 4), std::invalid_argument)
```
Comment (kparichay) [inline nntrainer/tensor/tensor.cpp:760]: add a throw for strided tensors as well.

File: nntrainer/tensor/tensor.cpp [hunk 3]
```diff
@@ -750,6 +752,49 @@ Tensor Tensor::getSharedDataTensor(const TensorDim dim_, unsigned int offset,
...
+    << "axis is not divisible by num_size, axis: " << axis
+    << " num size: " << num_size;
+
+  auto ret_dim = dim;
+  auto new_dim = dim.getTensorDim(axis) / num_size;
+  ret_dim.setTensorDim(axis, new_dim);
+
+  auto iter_value = [this, &ret_dim](std::array<unsigned, 4> &loc) {
+    auto value = getValue(loc[0], loc[1], loc[2], loc[3]);
+    for (int i = 3; i >= 0; --i) {
```
Comment (kparichay) [inline nntrainer/tensor/tensor.cpp:773]: I would highly recommend you to refer the split layer implementation, where changes the shape the tensor to be split to not only simplify the computation but also makes it quite efficient.

Comment (kparichay) [reply nntrainer/tensor/tensor.cpp:773]: this is return 1 value at a time with this operation, using getValue() - this is needed only along the axis for the split right and not along all the dimensions?

File: nntrainer/tensor/tensor.cpp [hunk 4]
```diff
@@ -750,6 +752,49 @@ Tensor Tensor::getSharedDataTensor(const TensorDim dim_, unsigned int offset,
...
+
+  NNTR_THROW_IF(dim.getTensorDim(axis) % num_size != 0, std::invalid_argument)
+    << "axis is not divisible by num_size, axis: " << axis
+    << " num size: " << num_size;
+
+  auto ret_dim = dim;
+  auto new_dim = dim.getTensorDim(axis) / num_size;
+  ret_dim.setTensorDim(axis, new_dim);
+
+  auto iter_value = [this, &ret_dim](std::array<unsigned, 4> &loc) {
```
Comment (kparichay) [inline nntrainer/tensor/tensor.cpp]: this return 1 value at a time with this operation, using getValue() - this is needed only along the axis for the split right and not along all the dimensions?

File: nntrainer/tensor/tensor.cpp [hunk 5]
```diff
@@ -750,6 +752,49 @@ Tensor Tensor::getSharedDataTensor(const TensorDim dim_, unsigned int offset,
   return ret;
 }
 
+std::vector<Tensor> Tensor::split(unsigned num_size, int axis) {
```
Comment (kparichay) [inline nntrainer/tensor/tensor.cpp:756]: throw if num_size == 0?

File: nntrainer/tensor/tensor.cpp [hunk 6]
```diff
@@ -795,6 +796,61 @@ std::vector<Tensor> Tensor::split(unsigned num_size, int axis) {
...
 }
 
+Tensor Tensor::cat(const std::vector<Tensor> &tensors, int axis) {
+  if (axis == -1) {
+    axis = 3;
+  }
+  NNTR_THROW_IF(tensors.empty(), std::invalid_argument)
+    << "given tensor vector is empty";
+  auto ref_dim = tensors.front().getDim();
+  ref_dim.setTensorDim(axis, 1);
```
Comment (kparichay) [inline nntrainer/tensor/tensor.cpp:806]: ensure axis in range [0, 4)

File: nntrainer/tensor/tensor.cpp [hunk 7]
```diff
@@ -795,6 +796,61 @@ std::vector<Tensor> Tensor::split(unsigned num_size, int axis) {
...
+                               auto cur_dim = t.getDim();
+                               cur_dim.setTensorDim(axis, 1);
+                               return ref_dim == cur_dim;
+                             }),
+                std::invalid_argument)
+    << " all tensor must have the same dimension except for the axis, ref_dim: "
+    << ref_dim << " axis : " << axis;
+
+  auto axis_dim = 0u;
+  for (auto &t : tensors) {
```
Comment (kparichay) [inline nntrainer/tensor/tensor.cpp:818]: minor comment : use std::accumulate()?

File: nntrainer/tensor/tensor.cpp [hunk 8]
```diff
@@ -795,6 +796,61 @@ std::vector<Tensor> Tensor::split(unsigned num_size, int axis) {
...
+  auto ret_dim = ref_dim;
+  ret_dim.setTensorDim(axis, axis_dim);
+
+  auto ret = Tensor(ret_dim);
+
+  std::array<unsigned, 4> loc = {0, 0, 0, 0};
+  for (auto &t : tensors) {
+    std::array<unsigned, 4> start_loc = loc;
+    for (auto i = 0u, sz = t.size(); i < sz; ++i) {
+      iter_value(loc, start_loc, ret, t.getDim()) = t.getValue(i);
```
Comment (kparichay) [inline nntrainer/tensor/tensor.cpp:846]: I would highly recommend you to refer the concat layer implementation for higher efficiency.

File: nntrainer/tensor/tensor.cpp (hunk 2 above)
Comment (zhoonit) [reply nntrainer/tensor/tensor.cpp:760]: doesn't apply support strided tensors? if not I'll make a throw here

File: nntrainer/tensor/tensor.cpp (hunk 5 above)
Comment (zhoonit) [reply nntrainer/tensor/tensor.cpp:756]: Done

File: nntrainer/tensor/tensor.cpp (hunk 3 above)
Comment (zhoonit) [reply nntrainer/tensor/tensor.cpp:773]: The strategy is good, for now I'll stay with naiver implementation and improve later, as this is only aimed to be used for non-critical path (weight saving)
I have greater plan to implement a iterator something like numpy.nditer('external_loop') to generalize the behavior

Comment (kparichay) [reply nntrainer/tensor/tensor.cpp:773]: alright, but these methods should work without iterator too for non-strided tensors as they will be more efficient.

File: nntrainer/tensor/tensor.cpp (hunk 8 above)
Comment (zhoonit) [reply nntrainer/tensor/tensor.cpp:846]: I'd like to forward discussion to https://github.com/nnstreamer/nntrainer/pull/1792/commits/7d0d1624df8b11244955a6ec12564eb67eb94ba8#r770239923.

Concat layer implementation can't be directly used here as this implementation support different axis as well, 
eg) Tensor::cat( [3, 2, 4 ,5] , [3, 4, 4, 5], axis = 1 )  results is [3, 6, 4, 5]

File: nntrainer/tensor/tensor.cpp (hunk 2 above)
Comment (kparichay) [reply nntrainer/tensor/tensor.cpp:760]: yup, my bad, it will work.

**Review by kparichay — APPROVED** [review]
LGTM

File: nntrainer/tensor/tensor.cpp (hunk 8 above)
Comment (zhoonit) [reply nntrainer/tensor/tensor.cpp:846]: Resolved offline that this is supported. I'll check when I improve those :)

File: nntrainer/tensor/tensor.cpp (hunk 3 above)
Comment (zhoonit) [reply nntrainer/tensor/tensor.cpp:773]: Yeah, as per offline discussion, the iterator should cover the scenario with an option.

File: nntrainer/tensor/tensor.cpp (hunk 4 above)
Comment (zhoonit) [reply nntrainer/tensor/tensor.cpp]: yeah. That can be optimized.

File: nntrainer/tensor/tensor.cpp (hunk 6 above)
Comment (zhoonit) [reply nntrainer/tensor/tensor.cpp:817]: done

File: nntrainer/tensor/tensor.cpp (hunk 7 above)
Comment (zhoonit) [reply nntrainer/tensor/tensor.cpp]: done

**Review by lhs8928 — APPROVED** [review]
LGTM

**Review by jijoongmoon — APPROVED** [review]
LGTM

**Pull request merged**